node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,796 @@
+	client := &http.Client{
+		Timeout: time.Second * 5,
+	}
+
+	query := fmt.Sprintf("%s/ibc/core/channel/v1/channels/%s/ports/wasm.%s", w.lcdUrl, channelID, w.contractAddress)
+	connResp, err := client.Get(query)
+	if err != nil {
+		w.logger.Error("channel query failed", zap.String("query", query), zap.Error(err))
+		return "", fmt.Errorf("query failed: %w", err)
+	}
+	connBody, err := io.ReadAll(connResp.Body)
+	if err != nil {
+		return "", fmt.Errorf("read failed: %w", err)
+	}
+	connResp.Body.Close()
+
+	var result ibcChannelQueryResults
+	err = json.Unmarshal(connBody, &result)
+	if err != nil {
+		return "", fmt.Errorf("failed to unmarshal response: %s, error: %w", string(connBody), err)
+	}
+
+	if len(result.Channel.ConnectionHops) != 1 {
+		return "", fmt.Errorf("response contained %d hops when it should return exactly one, json: %s", len(result.Channel.ConnectionHops), string(connBody))
+	}
+
+	w.logger.Info("queried connection ID", zap.String("channelID", channelID), zap.String("connectionID", result.Channel.ConnectionHops[0]))
+	return result.Channel.ConnectionHops[0], nil
+}
+
+// parseEvent parses a wasm event. If it is from the desired contract and for the desired action, it returns an event. Otherwise, it returns nil.
+func parseEvent[T any](logger *zap.Logger, desiredContract string, desiredAction string, event gjson.Result) (*T, error) {
+	eventType := gjson.Get(event.String(), "type")
+	if eventType.String() != "wasm" {
+		return nil, nil
+	}
+
+	attrBytes, err := parseWasmAttributes(logger, desiredContract, desiredAction, event)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse attributes: %w", err)
+	}
+
+	if attrBytes == nil {
+		return nil, nil // This can be returned for different event types, so it is not an error.
+	}
+
+	evt := new(T)
+	if err := json.Unmarshal(attrBytes, evt); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal event attributes: %w", err)
+	}
+
+	return evt, nil
+}
+
+// parseWasmAttributes parses the attributes in a wasm event. If the contract and action match the desired values (or the desired values are not set)
+// the attributes are loaded into a byte array of the marshaled json. If the event is not of the desired type, it returns nil.
+func parseWasmAttributes(logger *zap.Logger, desiredContract string, desiredAction string, event gjson.Result) ([]byte, error) {
+	attributes := gjson.Get(event.String(), "attributes")
+	if !attributes.Exists() {
+		return nil, fmt.Errorf("message does not contain any attributes")
+	}
+
+	contractAddressSeen := false
+	actionSeen := false
+	attrs := make(map[string]string)
+	for _, attribute := range attributes.Array() {
+		if !attribute.IsObject() {
+			logger.Warn("event attribute is invalid", zap.String("attribute", attribute.String()))
+			continue
+		}
+		keyBase := gjson.Get(attribute.String(), "key")
```

Comment:
given there is no deep nesting here this could use a struct and json unmarshal and be more readable

Reply:
Reworked the parsing code.